src/shogun/kernel/GaussianARDKernel.cpp

```diff
@@ -122,21 +107,10 @@ SGMatrix<float64_t> CGaussianARDKernel::get_parameter_gradient(
 		{
 			for (index_t k=0; k<num_rhs; k++)
 			{
-				SGVector<float64_t> avec=
-					((CDenseFeatures<float64_t>*) lhs)->get_feature_vector(j);
-				SGVector<float64_t> bvec=
-					((CDenseFeatures<float64_t>*) rhs)->get_feature_vector(k);
-
-				REQUIRE(avec.vlen==bvec.vlen, "Number of right and left hand "
-						"features must be the same\n");
-
-				float64_t result=0;
+				float64_t result=distance(j,k);
```

Comment:
Why would that be a problem? The gradient method than just checks the enum and does compute different gradients?
Automatic would be good I agree, but thats for the future for now ;)

Anyway, I think this refactoring makes a lot of sense, check out GPy, they also do this, though they just distinguish two cases: normal and ARD. We should do the same via a flag. I think in fact we can use a boolean, too. The pairwise scaling thing is infeasible anyways, so no need to worry about it, sorry that I brought that up.

http://gpy.readthedocs.org/en/latest/tuto_GP_regression.html